### PR TITLE
Parser: allow on-demand member decl parsing for other nominal types.

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
Can you add -parse to scale_test and write some tests for this as we discussed?

**Comment by nkcsgexi** [comment]
@slavapestov sure, I'm looking into that.

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -186,7 +186,9 @@ void PersistentParserState::parseMembers(IterableDeclContext *IDC) {
   SourceFile &SF = *IDC->getDecl()->getDeclContext()->getParentSourceFile();
   assert(!SF.hasInterfaceHash() &&
     "Cannot delay parsing if we care about the interface hash.");
+  assert(SF.Kind != SourceFileKind::SIL && "cannot delay parsing SIL");
```
Comment (slavapestov) [inline lib/Parse/ParseDecl.cpp]: Why not?

Comment (nkcsgexi) [reply lib/Parse/ParseDecl.cpp]: We need to have a way to keep `SILParserState` around till delayed parsing, which requires non-trivial work. I don't think there're fundamental blockers there, just we need to keep track more stuff.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Parse/ParseDecl.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/Parse/ParseDecl.cpp]: Ok, fair enough. Can you add a comment somewhere to that effect? We don’t care about SIL parsing performance much anyway, I was just curious.

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci Please Test Source Compatibility

**Comment by nkcsgexi** [comment]
All tests passed. Merging

**Comment by dcci** [comment]
@nkcsgexi This seems to have broken a lldb test that runs only in the full run. Mind if we take a look together?

**Comment by dcci** [comment]
Trying to get a reduced testcase.

**Comment by nkcsgexi** [comment]
@dcci Thanks for working on a reduced test case! I think we specifically disabled delayed parsing for REPL input here: https://github.com/apple/swift/blob/master/lib/Parse/ParseDecl.cpp#L3372.

**Comment by dcci** [comment]
In the meanwhile, @vedantk / @fredriss there's a bug in the test suite, as this is a REPL inline test, and should be run as part of the "regular" PR cycle. [cc:ing you as you did some work in the last month on the area]

**Comment by dcci** [comment]
This doesn't reproduce on the REPL from the cmdline, it could be a bug in the SBAPI.

```
Davides-Mac-Pro:bin davide$ ./lldb --repl
Welcome to Swift version 4.2-dev (LLVM 7b8e4a2b59, Clang 61c26a2ab2, Swift 1e499bf5a9).
Type :help for assistance.
  1> class A {init(a: Int) {}}
  2> class B : A {let x: Int; init() { x = 5 + 5; super.init(a: x) } }
  3> B().x
$R0: Int = 10
  4> extension B : CustomStringConvertible { public var description:String { return "class B\(x) is a subclass of class A"} }
  5> B().description
$R1: String = "class B10 is a subclass of class A"
```

**Comment by dcci** [comment]
@vedantk thanks for the explanation!

**Pull request merged**